### DISCUSSION TITLE
alternate data streams, print data

### DISCUSSION THREAD

**Discussion by theflakes** [ColinFinck/ntfs#19] [Q&A] [answered]
Hello and thanks for this crate.

I'm working on a file forensic tool. I'm able to get all alternate data streams and their names. But I'm struggling figuring out how to get to the data for each. Below is as far as I've gotten. Just not sure how to reference the data part of this structure.

```
let stream = attribute.value(fs)?;
println!("{:?}", stream);
```
Output is: `Resident(NtfsResidentAttributeValue { data: [34, 116, 104, 105, 115, 32, 105, 115, 32, 101, 118, 105, 108, 34, 32, 13, 10], position: NtfsPosition(Some(3253758424)), stream_position: 0 })`

I apologize as I'm not a developer, just a hacker. Any tips would be greatly appreciated.

This is the tool I'm working on: https://github.com/theflakes/fmd

thanks

**Comment by ColinFinck** [comment, marked as answer]
Hi! The `NtfsAttributeValue` object you got via `attribute.value(fs)?` implements the `NtfsReadSeek` trait to read the actual attribute value or move the read cursor inside the data stream.
The ntfs-shell example uses it here: https://github.com/ColinFinck/ntfs/blob/9348d729bcefe0b9473d29b8d7972d5e4038c910/examples/ntfs-shell/main.rs#L528-L537

In fact, `NtfsReadSeek` is also implemented for the more specific attribute value structures (such as `NtfsResidentAttributeValue`).

BONUS: If you need an object that implements the `std::io`-compatible `Read` and `Seek` traits, you can use the `attach` method of `NtfsAttributeValue`, which returns such an object.
Due to the way `Read` and `Seek` are defined, this requires the returned object to mutably borrow `fs`. To use `fs` for something else again, you can either let the returned object go out of scope or use its `detach` method to retrieve the original `NtfsAttributeValue` again.

**Reply by theflakes** [reply]
Thank you, this really helped.
